packages/dapp/src/hooks/useKeysManager.ts

```diff
@@ -0,0 +1,212 @@
+import type { VerificationMethodReference } from '@windingtree/org.json-schema/types/org.json';
+import type { AnySchema } from '@windingtree/org.id-utils/dist/object';
+import type { GenericStateRecord } from '../store';
+import { useState, useCallback, useEffect } from 'react';
+import { useAppReducer } from '../store/reducer';
+
+import { object, uid } from '@windingtree/org.id-utils';
+import { org } from '@windingtree/org.json-schema';
+
+export const revocationReasonValues = object.getDeepValue(org, 'definitions.VerificationMethodRevocationReference.properties.reason.enum') as string[];
+export type RevocationReference = typeof revocationReasonValues[number]
+
+export const allowedKeysTypes: string[] = ['EcdsaSecp256k1VerificationKey2019'];
+
+export type KeyType =
+	typeof allowedKeysTypes[number];
+
+const keyRecordSchema: AnySchema = {
+  type: 'object',
+  properties: {
+    id: {
+      type: 'string'
+    },
+    type: {
+      type: 'string',
+      enum: allowedKeysTypes
+    },
+    publicKey: {
+      type: 'string'
+    },
+    tag: {
+      type: 'string'
+    },
+    note: {
+      type: 'string'
+    },
+    revocation: {
+      type: 'object',
+      properties:{
+        reason: {
+          type: 'string',
+          enum: revocationReasonValues
+        },
+        invalidityDate: {
+          type: 'string',
+          format: 'data-time'
+        },
+      },
+      required: ['reason', 'invalidityDate']
+    }
+  },
+  required: ['id', 'type', 'publicKey', 'tag']
+}
+
+export interface KeyRecord extends GenericStateRecord {
+  type: KeyType;
+	publicKey: string; // Ethereum account address
+	tag: string; // Unique across records formatted string
+	note?: string; // Key description
+	revocation?: VerificationMethodReference['verificationMethodRevocation']; // see this type for details
+}
+
+export type Keys = KeyRecord[];
+
+export type UseKeysManagerHook = [
+  addKey: Function,
+  updateKey: Function,
+  removeKey: Function,
+  revokeKey: Function,
+  loading: boolean,
+  error: string | undefined,
+];
+
+// UseKeysManager react hook
+export const useKeysManager = (): UseKeysManagerHook => {
+  const [state, dispatch] = useAppReducer();
```

Comment:
In one of my previous comments, I have proposed to use another declaration as:

```typescript
const { keys } = useAppState();
const dispatch = useAppDispatch();
```

at line 82 the dependency `[state.keys]` will not work properly and looks not obvious. Also when you import the whole state and use just one property without deconstruction a compiler will place the whole state to the scope.

Reply:
✔️